File: frontend/src/types/task.ts
export enum DurationType {
    DAYS = "days",
    WEEKS = "weeks",
    MONTHS = "months",
    YEARS = "years"
}

export interface Status {
    id: number;
    name: string;
    code: string;
    color: string;
    order: number;
    is_active: boolean;
    is_default: boolean;
    is_final: boolean;
}

export interface Priority {
    id: number;
    name: string;
    color: string;
    order: number;
    position?: number;
    is_active: boolean;
    is_default: boolean;
}

export interface Duration {
    id: number;
    name: string;
    type: DurationType;
    value: number;
    is_active: boolean;
    is_default: boolean;
    duration_type?: string;
}

export interface TaskType {
    id: number;
    name: string;
    description?: string;
    color?: string;
    order?: number;
    is_default?: boolean;
    is_active?: boolean;
}

export interface Task {
    id: number;
    title: string;
    description?: string;
    type?: TaskType;
    status?: Status;
    priority?: Priority;
    duration?: Duration;
    deadline?: string;
    deadline_iso?: string;
    created_at: string;
    completed_at?: string;
    completed?: boolean;
    is_overdue: boolean;
    reminders: string[];
    tags: string[];
    custom_fields: Record<string, unknown>;
}

export interface Settings {
    statuses: Status[];
    priorities: Priority[];
    durations: Duration[];
}

export interface CreateTaskDto {
    title: string;
    description?: string | null;
    type_id?: number;
    status_id?: number;
    priority_id?: number;
    duration_id?: number;
    deadline?: string;
    completed?: boolean;
}

export interface UpdateTaskDto {
    title?: string;
    description?: string | null;
    type_id?: number;
    status_id?: number;
    priority_id?: number;
    duration_id?: number;
    deadline?: string;
    completed?: boolean;
} 

File: backend/models/task.py
from sqlalchemy import Column, Integer, String, Text, DateTime, ForeignKey, JSON, Boolean
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from backend.database import Base

class TaskTypeSetting(Base):
    __tablename__ = "task_type_settings"

    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(String, index=True)
    name = Column(String)
    description = Column(Text, nullable=True)
    color = Column(String, nullable=True)
    order = Column(Integer, default=0)
    is_default = Column(Boolean, default=False)
    is_active = Column(Boolean, default=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())

    tasks = relationship("Task", back_populates="type")

class Task(Base):
    __tablename__ = "tasks"

    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(String, index=True)
    title = Column(String)
    description = Column(Text, nullable=True)
    type_id = Column(Integer, ForeignKey("task_type_settings.id"), nullable=True)
    status_id = Column(Integer, ForeignKey("statuses.id"), nullable=True)
    priority_id = Column(Integer, ForeignKey("priorities.id"), nullable=True)
    duration_id = Column(Integer, ForeignKey("durations.id"), nullable=True)
    deadline = Column(DateTime(timezone=True), nullable=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())

    type = relationship("TaskTypeSetting", back_populates="tasks")
    status = relationship("Status", back_populates="tasks")
    priority = relationship("Priority", back_populates="tasks")
    duration = relationship("Duration", back_populates="tasks") 